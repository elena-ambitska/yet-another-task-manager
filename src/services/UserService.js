import BaseService from "./BaseService";

class UserService extends BaseService {

    async register(data) {
        return this.post(`auth/local/register`, data);
    }

    async login(data) {
        return this.post(`auth/local`, data);
    }
}

export default new UserService();