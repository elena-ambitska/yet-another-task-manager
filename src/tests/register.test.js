import React from "react";
import Register from "../components/register";
import {shallow} from "enzyme";
import toJson from "enzyme-to-json";
import UserService from "../services/UserService";
import registerMock from "./mock/register";
import registerFailMock from "./mock/register_fail_400";


const getUserServiceDataMock = jest.spyOn(UserService, "register");

beforeEach(() => {
    getUserServiceDataMock.mockClear();
});

const mockedUsedNavigate = jest.fn();

jest.mock('react-router', () => ({
    ...jest.requireActual('react-router'),
    useNavigate: () => mockedUsedNavigate,
}));

describe("Test Register component", () => {

    it("should render register component", () => {
        const component = shallow(<Register/>);
        expect(toJson(component)).toMatchSnapshot();
    });

    it("register component sends data to services", async () => {

        getUserServiceDataMock.mockReturnValueOnce(Promise.resolve(registerMock));

        const component = shallow(<Register/>);
        await component
            .find("form")
            .simulate("submit", {
                preventDefault() {
                }
            });

        expect(getUserServiceDataMock).toHaveBeenCalledTimes(1);

    });

    it("register component displays error when fails", async () => {

        getUserServiceDataMock.mockReturnValueOnce(Promise.resolve(registerFailMock));

        const component = shallow(<Register/>);
        await component
            .find("form")
            .simulate("submit", {
                preventDefault() {
                }
            });

        expect(getUserServiceDataMock).toHaveBeenCalledTimes(1);

        const errorList = component.find(".errors-list");
        expect(errorList.children.length).toBe(registerFailMock.data[0].messages.length);
        expect(errorList.childAt(0).text()).toContain(registerFailMock.data[0].messages[0].message);

    });

    it("register component displays password errors", async () => {


        getUserServiceDataMock.mockReturnValueOnce(Promise.resolve(registerFailMock));

        const component = shallow(<Register/>);
        const input = component.find("input[name='password']");
        input.props();
        input.simulate('change', {
            preventDefault() {
            },
            target: {
                name: 'password',
                value: 'pass 123',
            }
        });

        await component
            .find("form")
            .simulate("submit", {
                preventDefault() {
                }
            });

        expect(getUserServiceDataMock).toHaveBeenCalledTimes(0);

        const errorList = component.find(".errors-list");
        expect(errorList.children.length).toBe(registerFailMock.data[0].messages.length);
        expect(errorList.childAt(0).text()).toContain("Error passwords not match");
    });
});
