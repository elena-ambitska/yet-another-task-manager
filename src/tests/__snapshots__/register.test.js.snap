// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Register component should render register component 1`] = `
<article
  className="grid"
>
  <form
    onSubmit={[Function]}
  >
    <h2>
      Register
    </h2>
    <div
      className="mb-3"
    >
      <label
        className="form-label"
        htmlFor="nameInput"
      >
         Name
      </label>
      <input
        className="form-control"
        id="nameInput"
        name="login"
        onChange={[Function]}
        required={true}
        value=""
      />
    </div>
    <div
      className="mb-3"
    >
      <label
        className="form-label"
        htmlFor="emailInput"
      >
         Email
      </label>
      <input
        className="form-control"
        id="emailInput"
        name="email"
        onChange={[Function]}
        required={true}
        type="email"
        value=""
      />
    </div>
    <div
      className="mb-3"
    >
      <label
        className="form-label"
        htmlFor="passwordInput"
      >
        Password
      </label>
      <input
        className="form-control"
        id="passwordInput"
        name="password"
        onChange={[Function]}
        required={true}
        type="password"
        value=""
      />
    </div>
    <div
      className="mb-3"
    >
      <label
        className="form-label"
        htmlFor="passwordConfirmInput"
      >
         Password Confirm
      </label>
      <input
        className="form-control"
        id="passwordConfirmInput"
        name="retype"
        onChange={[Function]}
        required={true}
        type="password"
        value=""
      />
    </div>
    <button
      className="btn btn-primary"
    >
      Register
    </button>
  </form>
  <ul
    className="errors-list"
  />
</article>
`;
